File: 2331-evaluate-boolean-binary-tree/2331-evaluate-boolean-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool postOrder(TreeNode*t){
        
        if(!t->left && !t->right)return t->val;
        
        bool left=postOrder(t->left);
        bool right=postOrder(t->right);
        if(t->val==2)
            return left||right;
        return left && right;
        
    }
    
    bool evaluateTree(TreeNode* root) {
        return postOrder(root);
    }
};

File: 2331-evaluate-boolean-binary-tree/2331-evaluate-boolean-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public boolean postOrder(TreeNode t){
        if(t.val==0 )return false;
        else if(t.val==1)return true;
        else if(t.val==2)return postOrder(t.left)||postOrder(t.right);
        return postOrder(t.left)&& postOrder(t.right);
    }
    
    public boolean evaluateTree(TreeNode root) {
        return postOrder(root);
    }
}